src/pygeoprocessing/geoprocessing.py

```diff
@@ -2495,37 +2487,48 @@ def convolve_2d(
             kernel_block[numpy.isclose(kernel_block, kernel_nodata)] = 0.0
         kernel_sum += numpy.sum(kernel_block)
 
+    work_queue = queue.Queue()
+    signal_offset_list = list(iterblocks(s_path_band, offset_only=True))
+    kernel_offset_list = list(iterblocks(k_path_band, offset_only=True))
+    n_blocks = len(signal_offset_list) * len(kernel_offset_list)
+
+    LOGGER.debug(f'start fill work queue thread')
+
+    def _fill_work_queue():
+        """Asynchronously fill the work queue."""
+        LOGGER.debug('fill work queue')
+        for signal_offset in signal_offset_list:
+            for kernel_offset in kernel_offset_list:
+                work_queue.put((signal_offset, kernel_offset))
+        work_queue.put(None)
+        LOGGER.debug('work queue full')
+
+    fill_work_queue_worker = threading.Thread(
+        target=_fill_work_queue)
+    fill_work_queue_worker.daemon = True
+    fill_work_queue_worker.start()
+
     # limit the size of the write queue so we don't accidentally load a whole
     # array into memory, work queue is okay because it's only passing block
     # indexes
-    work_queue = queue.Queue()
-    write_queue = queue.Queue(n_threads * 2)
-
-    # workers is n_threads - 1 because we count the current thread
+    LOGGER.debug(f'start worker thread')
```

Comment:
another `f` here with no string replacement.

Reply:
Removed in 79409027d7d280b665fd2812aa4437350acfa59b